File: 初始算法/链表/反转链表.java
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode(int x) { val = x; } }
 */

class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
    }
}

class Solution {
    private ListNode node;

    public ListNode reverseList(ListNode head) {
        _reverse(head);
        return node;
    }

    private ListNode _reverse(ListNode head) {
        if (head == null) {
            return null;
        }
        ListNode reverseHead = _reverse(head.next);
        if (reverseHead != null) {
            reverseHead.next = head;
            head.next = null;
        } else {
            // 这里的head是我们需要的真正的头指针
            node = head;
        }
        return head;
    }
}

File: 数据结构/链表/经典问题/反转链表.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
#include <iostream>
using namespace std;
struct ListNode {
     int val;
     ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(!head || !(head->next)) {
            return head;
        }
        
        
        ListNode *pre = NULL;
        ListNode *cur = head;
        ListNode *next = head->next;
        
        
        while(cur) {
            cur->next = pre;
            pre = cur;
            cur = next;
            if(next) {
                next = next->next;
            }
        }
        
        
        return pre;
    }
};